lighthouse-logger/index.js

```diff
@@ -118,33 +166,153 @@ class Log {
+    Log[level]('statusEnd', msg, ...(args || []));
     marky.stop(id);
   }
 
-  static log(title, ...args) {
-    Log.events.issueStatus(title, args);
-    return Log._logToStdErr(title, args);
+  /* eslint-disable no-invalid-this */
+  /**
+   * Decorates a function, calling time/timeEnd before/after calling the original function.
+   * @template T
+   * @template {*} R
+   * @template {*[]} Args
+   * @param {(this: T, ...args: Args) => R} originalFn
+   * @param {TimeDecorateOpts<T, Args>} opts
+   * @return {(this: T, ...args: Args) => R}
+   */
+  static timeDecorate(originalFn, opts) {
+    /**
+     * @type {(this: *, ...args: *[]) => string}
+     */
+    const computeMsg = (_this, args) => {
+      if (typeof opts.msg === 'string') return opts.msg;
+      if (typeof opts.msg === 'function') return opts.msg.apply(_this, args);
+      throw new Error('expected msg');
+    };
+
+    /**
+     * @type {(this: *, ...args: *[]) => string}
+     */
+    const computeId = (_this, args) => {
+      if (typeof opts.id === 'string') return opts.id;
+      if (typeof opts.id === 'function') return opts.id.apply(_this, args);
+      return `lh:${originalFn.name}`;
+    };
+
+    /**
+     * @type {(this: T, ...args: Args) => R}
+     */
+    const fn = function timeDecoratedFn(...args) {
+      const timeStartLogLevel = opts.timeStartLogLevel || 'log';
+      const timeEndLogLevel = opts.timeEndLogLevel || 'verbose';
+
+      const status = {msg: computeMsg(this, args), id: computeId(this, args)};
+      Log.time(status, timeStartLogLevel);
+
+      let result;
+      try {
+        result = originalFn.apply(this, args);
+      } catch (err) {
+        Log.timeEnd(status, timeEndLogLevel);
+        // intercept any errors and elide the time decoration from the stack trace
+        err.stack = err.stack.replace(/.* at Function\.timeDecoratedFn .*\n/g, '');
+        throw err;
+      }
+
+      if (result && typeof result.then === 'function') {
+        return result.then((/** @type {any} */ value) => {
+          Log.timeEnd(status, timeEndLogLevel);
+          return value;
+        }).catch((/** @type {any} */ err) => {
+          Log.timeEnd(status, timeEndLogLevel);
+          // intercept any errors and elide the time decoration from the stack trace
+          err.stack = err.stack.replace(/.* at Function\.timeDecoratedFn .*\n/, '');
+          throw err;
+        });
+      } else {
+        Log.timeEnd(status, timeEndLogLevel);
+        return result;
+      }
+    };
+    return fn;
   }
+  /* eslint-enable no-invalid-this */
 
-  static warn(title, ...args) {
-    Log.events.issueWarning(title, args);
-    return Log._logToStdErr(`${title}:warn`, args);
+  /**
+   * Decorates (like timeDecorate) specified methods of a class.
+   * If decorating instance methods, use the class's prototype.
+   * If decorating static methods, use the class directly.
+   * @template {Object|Function} Class
+   * @template {keyof Class} Prop
```

Comment:
should be limited to properties that are functions? E.g. you can pass a number property in (and it will accept any parameters for `msg` and `id`)

Reply:
this part guards against non-functions:

```
* @param {{[key in Prop]: TimeDecorateOpts<Class, Parameters<IsFunction<Class[key]>>>}} methods
```

`IsFunction<T>` resolves to never if `T` does not extend `Function`.